tests/test_packages.py

```diff
@@ -264,41 +269,78 @@ def test_LocalPackageResolver_lookup_packages(dep_name, expect_import_names):
+                )
+            },
             {
                 "apache_airflow": Package(
                     "apache-airflow", {DependenciesMapping.USER_DEFINED: {"airflow"}}
                 ),
                 "pip": Package("pip", {DependenciesMapping.LOCAL_ENV: {"pip"}}),
                 "pandas": Package("pandas", {DependenciesMapping.IDENTITY: {"pandas"}}),
             },
-            id="mixed_deps__uses_mixture_of_user_defined_identity_and_local_env_mapping",
+            id="mixed_deps__uses_mixture_of_user_defined_from_file_identity_and_local_env_mapping",
         ),
         pytest.param(
             ["pandas", "pip"],
-            dedent("""apache-airflow = ["airflow"]"""),
+            {"file": dedent("""apache-airflow = ["airflow"]""")},
             {
                 "pip": Package("pip", {DependenciesMapping.LOCAL_ENV: {"pip"}}),
                 "pandas": Package("pandas", {DependenciesMapping.IDENTITY: {"pandas"}}),
             },
             id="mixed_deps__unaffected_by_nonmatching_user_defined_mapping",
         ),
+        pytest.param(
+            ["pandas", "pip", "apache_airflow"],
+            {"configuration": {"apache-airflow": ["airflow"]}},
+            {
+                "apache_airflow": Package(
+                    "apache-airflow", {DependenciesMapping.USER_DEFINED: {"airflow"}}
+                ),
+                "pip": Package("pip", {DependenciesMapping.LOCAL_ENV: {"pip"}}),
+                "pandas": Package("pandas", {DependenciesMapping.IDENTITY: {"pandas"}}),
+            },
+            id="mixed_deps__user_defined_from_config_identity_and_local_env_mapping",
+        ),
+        pytest.param(
+            ["pandas", "pip", "apache_airflow"],
+            {
+                "file": dedent("""apache-airflow = ["airflow"]"""),
+                "configuration": {"apache-airflow": ["foo", "bar"]},
+            },
+            {
+                "apache_airflow": Package(
+                    "apache-airflow",
+                    {DependenciesMapping.USER_DEFINED: {"airflow", "foo", "bar"}},
+                ),
+                "pip": Package("pip", {DependenciesMapping.LOCAL_ENV: {"pip"}}),
+                "pandas": Package("pandas", {DependenciesMapping.IDENTITY: {"pandas"}}),
+            },
+            id="mixed_deps__uses_mixture_of_user_defined_identity_and_local_env_mapping",
+        ),
     ],
 )
 def test_resolve_dependencies__focus_on_mappings(
     dep_names, user_mapping, expected, write_tmp_files
 ):
-    user_mapping_path = None
+    custom_mapping_file = None
+    custom_mapping = None
     if user_mapping is not None:
-        tmp_path = write_tmp_files({"mapping.toml": user_mapping})
-        user_mapping_path = tmp_path / "mapping.toml"
+        custom_mapping = user_mapping.get("configuration")
+        if "file" in user_mapping:
+            tmp_path = write_tmp_files({"mapping.toml": user_mapping["file"]})
+            custom_mapping_file = tmp_path / "mapping.toml"
```

Comment:
In this specific use case of writing a single file I don't think `write_tmp_files()` adds much value. Consider using `tmp_path` directly instead:
```suggestion
            custom_mapping_file = tmp_path / "mapping.toml"
            custom_mapping_file.write_text(dedent(user_mapping["file"]))
```
Then, replace `write_tmp_files` with `tmp_path` in the argument list, and drop the `dedent()` calls in the parametrization above.